### ISSUE TITLE
split the single gigantic FCS into multiple files

### ISSUE THREAD

**Issue by mikejiang** [RGLab/flowCore#74]
The issue is described below and the best way to handle this is to provide an API to split the file so that they can be loaded back as disk-based `ncdfFlowSet` because we probably don't want to read the entire data into a single `flowFrame` object.

On 05/23/2017 03:51 PM, Mike wrote:
>
> Vinko,
>
> Thank you for sharing the data. If you look at the header
>
> > as.numeric(as.list(read.FCSheader(file)[[1]])[["$ENDDATA"]])
> [1] 5857291721
> > .Machine$integer.max
> [1] 2147483647
>
> You will find the number of bytes in data section exceeds the R’s integer limit, which fails readBin call.
>
>>> From: "vinko@cytometry.uzh.ch" <vinko@cytometry.uzh.ch>
>>> Date: Monday, May 22, 2017 at 1:46 PM
>>> To: "Finak, Greg" <gfinak@fredhutch.org>
>>> Subject: Re: Bioconductor Flow tools 
>>>
>>>  
>>>
>>> BTW, speaking of flowCore - what's the best place to look for help? One of my users somehow generated 5GB file on the BD instrument, containing 90 Mio events! Now he can't open it anywhere. I thought it shouldn't be a problem for flowCore to read it, but it returns an error I never saw before:
>>>
>>>
>>> Error in readBin(con = con, what = dattype, n = count, size = size, signed = signed,  :
>>>   vector size cannot be NA/NaN
>>> In addition: Warning message:
>>> In readFCSdata(con, offsets, txt, transformation, which.lines, scale,  :
>>>   NAs introduced by coercion to integer range
>>>
>>>
>>>
>>>

**Comment by SamGG** [comment]
longlong seems to be an implementation/compilation choice. My guess about 64 bits is wrong ;-)
```
> .Machine$sizeof.longlong
[1] 8
```
but ```help(integer)``` states ```Note that current implementations of R use 32-bit integers for integer vectors```

**Comment by mikejiang** [comment]
I added the boundary check so that it gives more meaningful message.
```r
> fr <- read.FCS(file)
 Error in readFCSdata(con, offsets, txt, transformation, which.lines, scale,  : 
  Total number of bytes (5857286216) in data segment exceeds the R integer limits!Please read the subset of FCS by specifying 'which.lines'
```
As it indicates, the workaround would be specifying row indices through `which.lines` argument so that it bypass the whole block reading
```r
nTotal <- as.numeric(as.list(read.FCSheader(file)[[1]])[["$TOT"]])
 nTotal
> nTotal
[1] 86136562
  fr <- read.FCS(file, which.lines = seq_len(nTotal))
```
This would be slow since it performs disk IO for each row.

**Comment by SamGG** [comment]
Just being curious. Skipping 2^31 lines does not help?

**Comment by mikejiang** [comment]
Not sure if I understand your question

**Comment by SamGG** [comment]
Sorry. I thought there was a skip.lines option in read.FCS. I mistake with another read function.

**Comment by dtelad11** [comment]
I submitted a pull request with a proposed solution:

https://github.com/RGLab/flowCore/pull/95

I managed to import a 14.4m event file on my desktop computer.

**Comment by gfinak** [comment]
@vtosevski would you be willing to post the file somewhere privately, we could do some benchmarking?

**Comment by dtelad11** [comment]
@vtosevski do not use the `which.lines` argument. When you use it, `read.FCS` seeks and reads one line at a time. This is a very expensive operation that takes several seconds per line. I doubt the 9m lines you're trying to read will finish during your lifetime.

The solution I proposed in #95 has been merged into the RGLab version of flowCore. You should be able to load the files by just calling read.FCS, without any parameters. 64GB should be more than enough to import the file you're describing.

**Comment by vtosevski** [comment]
@dtelad11 thanks David! I haven't followed through #95 but was working with Mike's comment from above in this thread. How do I get that "RGLab version of flowCore" on my workstation? I currently run 1.44.2 from Bioconductor.

**Comment by gfinak** [comment]
Install the devtools package and then do
`devtools::install_github("RGLab/flowCore",ref="trunk")`
You may need to install other dependencies as well.

**Comment by vtosevski** [comment]
@gfinak thanks for the instructions! I have just installed the trunk branch version and read in the previously mentioned 18 Mio events file. It worked like a charm, I am amazed!
```
   user  system elapsed 
  39.33    4.90   44.39 
```
Thanks Greg and David! @dtelad11

**Comment by dtelad11** [comment]
Great, I'm glad to hear it worked out. I've been using it with some large
files and it's encouraging to hear it's working for others as well.

**Comment by gfinak** [comment]
Can this be closed?

**Comment by vtosevski** [comment]
Yes, absolutely. El-ad's solution worked like a charm.

Thanks everyone!

Vinko

**Comment by mikejiang** [comment]
solved by PR #95
